### ISSUE TITLE
Refinement dropped when specializing type class constraint

### ISSUE THREAD

**Issue by Gabriel439** [ucsd-progsys/liquidhaskell#602]
The following program incorrectly type-checks:

```haskell
{-@
class Foo a where
    foo :: { x : a | x /= 0 } -> a
@-}

class Foo a where
    foo :: a -> a

instance Foo Double where
    foo = id

-- example :: (Num a, Foo a) => a
example :: Double
example = foo 0
```

If you use the commented out type signature for `example` then type-checking correctly fails.

I tested this against a freshly checked out repository following the `stack` build instructions.

**Comment by gridaphobe** [comment]
Thanks for the minimal example, this is clearly bogus!

**Comment by ranjitjhala** [comment]
Yikes! This is an alarming soundness bug. @nikivazou or @gridaphobe can you take a look?

(Thanks for pointing it out @Gabriel439!)

**Comment by gridaphobe** [comment]
Sorry for the delay @Gabriel439! It looks like the problem is that the `0` in the refined class signature is not overloaded the way a Haskell literal is.

@nikivazou can you take over? I'm not very familiar with the int/real handling. I just pushed a `T602` branch that includes a few related fixes for handling classes in general. (In particular, see https://github.com/ucsd-progsys/liquidhaskell/blob/bae9b39b00778a4b11431ce01b9bfc5d941f95cc/tests/neg/T602.hs)

**Comment by nikivazou** [comment]
That seems quite tricky.... 

The problem is that the precondition for `foo` is pruned out as unsorted. 
The following two programs are Unsafe 

```
foo :: { x : a | x /= 0.0 } -> a
example :: Double
```

and 
```
foo :: { x : a | x /= 0 } -> a
example :: (Num a, Foo a) => a
```


But the combination 
```
foo :: { x : a | x /= 0 } -> a
example :: Double
```
 makes foo's precondition unsorted and prunes it out. 
Not sure what is a good solution to this, other than disabling pruning...

**Comment by ranjitjhala** [comment]
Oh no. Not the pruning again...

Why are we pruning in the specification? This *must* be an error reported to the user.

**Comment by gridaphobe** [comment]
This was fixed by #650 (in `develop` but not in `master` yet).
